theano/tensor/sort.py

```diff
@@ -206,3 +218,275 @@ def argsort(a, axis=-1, kind='quicksort', order=None):
         a = a.flatten()
         axis = 0
     return ArgSortOp(kind, order)(a, axis)
+
+
+if hasattr(np, 'argpartition'):
+    # numpy >= 1.8 implementation
+    def _topk_py_impl(op, x, k, axis, idx_dtype):
+        ndim = x.ndim
+        if abs(k) == 1:
+            i = (k + 1) // 2
+            fn_max = [np.min, np.max][i]
+            fn_argmax = [np.argmin, np.argmax][i]
+            if not op.return_indices:
+                return np.expand_dims(fn_max(x, axis=axis), axis)
+            elif op.return_values:
+                zi = np.expand_dims(
+                    fn_argmax(x, axis=axis).astype(idx_dtype), axis)
+                idx2 = tuple(
+                    np.arange(s).reshape(
+                        (s,) + (1,) * (ndim - i - 1)
+                        ) if i != axis else zi for i, s in enumerate(x.shape))
+                zv = x[idx2]
+                return zv, zi.astype(idx_dtype)
+            else:
+                zi = np.expand_dims(
+                    fn_argmax(x, axis=axis).astype(idx_dtype), axis)
+                return zi.astype(idx_dtype)
+
+        asize = x.shape[axis]
+        if asize == abs(k):
+            if not op.return_indices:
+                return x.copy()
+            else:
+                l = axis
+                r = ndim - l
+                reps = list(x.shape)
+                reps[axis] = 1
+                zi = np.arange(abs(k), dtype=idx_dtype)
+                zi = zi.reshape((1,) * l + (k,) + (1,) * (r - 1))
+                zi = np.tile(zi, reps)
+                if op.return_values:
+                    return x.copy(), zi
+                else:
+                    return zi
+
+        idx = [slice(None)] * ndim
+        if k > 0:
+            idx[axis] = slice(-k, None)
+        elif k < 0:
+            idx[axis] = slice(-k)
+        else:
+            raise ValueError('k cannot be zero')
+        if not op.return_indices:
+            zv = np.partition(x, -k, axis=axis)[idx]
+            return zv
+        elif op.return_values:
+            zi = np.argpartition(x, -k, axis=axis)[idx]
+            idx2 = tuple(
+                np.arange(s).reshape(
+                    (s,) + (1,) * (ndim - i - 1)
+                    ) if i != axis else zi for i, s in enumerate(x.shape))
+            zv = x[idx2]
+            return zv, zi.astype(idx_dtype)
```

Comment:
I just did the timing, although not extensive, for a `4096 x 4096` `float64` array, finding the first 20 largest values and indices. Taking indices is almost 2x fast than sorting twice. The timing includes theano functions call overhead.

Taking indices : **0.165s**
Sorting twice : **0.318s**

The result is averaged from 10 calls, excluding the 1st call.